internal/runtime/client/client.go

```diff
@@ -153,6 +157,106 @@ func (c *client) Unregister(extensionConfig *runtimev1.ExtensionConfig) error {
+		hook:   hook,
+	}
+}
+
+// HookClient is the client used to make calls of the runtime hooks and runtime extension handlers.
+type HookClient interface {
+	// CallAll calls all the extension registered for the hook.
+	CallAll(ctx context.Context, request runtime.Object, response runtimehooksv1.AggregatableResponse) error
+
+	// Call calls only the extension with the given name.
+	Call(ctx context.Context, name string, request, response runtime.Object) error
+}
+
+type hookClient struct {
+	client *client
+	hook   catalog.Hook
+}
+
+func (h *hookClient) CallAll(ctx context.Context, request runtime.Object, response runtimehooksv1.AggregatableResponse) error {
+	gvh, err := h.client.catalog.GroupVersionHook(h.hook)
+	if err != nil {
+		return errors.Wrap(err, "failed to compute GroupVersionHook")
+	}
+	registrations, err := h.client.registry.List(catalog.GroupHook{Group: gvh.Group, Hook: gvh.Hook})
+	if err != nil {
+		return errors.Wrap(err, "failed to retrieve ExtensionHandlers information")
+	}
+	responses := []*runtimehooksv1.ExtensionHandlerResponse{}
+	// Future-work: Call the extension handlers concurrently.
+	for _, registration := range registrations {
+		tmpResponse, err := h.client.catalog.NewResponse(gvh)
+		if err != nil {
+			return errors.Wrap(err, "failed to create respons object")
+		}
+		err = h.Call(ctx, registration.Name, request, tmpResponse)
+		// If at least once of the extension handlers failed lets short-circuit here and return early.
+		if err != nil {
+			return errors.Wrapf(err, "ExtensionHandler %s failed", registration.Name)
+		}
+		responses = append(responses, &runtimehooksv1.ExtensionHandlerResponse{
+			Name:     registration.Name,
+			Response: tmpResponse,
+		})
+	}
+	if err := response.Aggregate(responses); err != nil {
+		return errors.Wrap(err, "failed to aggregate responses")
+	}
+	return nil
+}
+
+func (h *hookClient) Call(ctx context.Context, name string, request, response runtime.Object) error {
+	registration, err := h.client.registry.Get(name)
+	if err != nil {
+		return err
+	}
+	var timeoutDuration time.Duration
+	if registration.TimeoutSeconds != nil {
+		timeoutDuration = time.Duration(*registration.TimeoutSeconds) * time.Second
+	}
+	opts := &httpCallOptions{
+		catalog: h.client.catalog,
+		config:  registration.ClientConfig,
+		gvh:     registration.GroupVersionHook,
+		name:    strings.TrimSuffix(registration.Name, "."+registration.ExtensionConfigName),
+		timeout: timeoutDuration,
+	}
+	err = httpCall(ctx, request, response, opts)
+	if err != nil {
+		// If the error is errCallingExtensionHandler then apply failure policy to calculate
+		// the effective result of the operation.
+		ignore := *registration.FailurePolicy == runtimev1.FailurePolicyIgnore
+		if _, ok := err.(*errCallingExtensionHandler); ok && ignore {
+			// Update the response to a default success response and return.
+			if err := runtimehooksv1.SetStatus(response, runtimehooksv1.ResponseStatusSuccess); err != nil {
```

Comment:
I wonder if we should also set the message to empty at this point, because it is supposed to contain an error message